lazy_tensor_core/lazy_tensor_core/csrc/LazyShapeInference.cpp

```diff
@@ -45,15 +45,70 @@
 
 // #include <functional>
 #include "torch/csrc/lazy/core/shape.h"
+#include <ATen/native/ConvUtils.h>
 #include "aten/src/ATen/native/ReduceOpsUtils.h"
 #include "lazy_tensor_core/csrc/ts_backend/LazyShapeInference.h"
 #include "torch/csrc/api/include/torch/enum.h"
+#include <iostream>
+#include <vector>
 
 namespace torch_lazy_tensors {
 namespace ir {
 namespace ops {
 using Shape = torch::lazy::Shape;
 
+
+// Copied from ATen/native/utils/ParamUtils.h, which aparently I can't include from here?
+std::vector<int64_t> expand_param_if_needed(
+    at::IntArrayRef list_param,
+    const char* param_name,
+    int64_t expected_dim) {
+  if (list_param.size() == 1) {
+    return std::vector<int64_t>(expected_dim, list_param[0]);
+  } else if ((int64_t)list_param.size() != expected_dim) {
+    std::ostringstream ss;
+    ss << "expected " << param_name << " to be a single integer value or a "
+       << "list of " << expected_dim << " values to match the convolution "
+       << "dimensions, but got " << param_name << "=" << list_param;
+    AT_ERROR(ss.str());
+  } else {
+    return list_param.vec();
+  }
+}
+
+std::vector<Shape> compute_shape_convolution_backward(const at::Tensor & grad_output, const at::Tensor & input, const at::Tensor & weight, c10::optional<at::IntArrayRef> bias_sizes, at::IntArrayRef stride, at::IntArrayRef padding, at::IntArrayRef dilation, bool transposed, at::IntArrayRef output_padding, int64_t groups, ::std::array<bool,3> output_mask) {
+  if (bias_sizes.has_value()) {
+    return {Shape(input.scalar_type(), input.sizes().vec()),
+            Shape(weight.scalar_type(), weight.sizes().vec()),
+            Shape(grad_output.scalar_type(), bias_sizes.value().vec())};
+  } else {
+    // TODO(whc) not sure whether to return 2 shapes here, or a 3rd one that is empty
+    return {Shape(input.scalar_type(), input.sizes().vec()),
+            Shape(weight.scalar_type(), weight.sizes().vec())};
+  }
+}
+
+std::vector<Shape> compute_shape_convolution(const at::Tensor & input, const at::Tensor & weight, const c10::optional<at::Tensor> & bias, at::IntArrayRef stride, at::IntArrayRef padding, at::IntArrayRef dilation, bool transposed, at::IntArrayRef output_padding, int64_t groups) {
+
+  int64_t dim = weight.ndimension() - 2;
```

Comment:
2 here means minus batch size dimension and minus channel dim? Basically, the spatial dimensions left, so we expand strides, padding to the size of spatial dimensions?

Reply:
yep.  And this is borrowed from some native code so i know its consistent